Remotion/Web/ClientScript/Scripts/SmartPage.ts

```diff
@@ -0,0 +1,1307 @@
+      return e.target as HTMLElement;
+    else if (TypeUtility.IsDefined(e.srcElement) && e.srcElement != null)
+      return e.srcElement as HTMLElement;
+    else
+      return null;
+  };
+
+  private GetPageRequestManager(): Nullable<Sys.WebForms.PageRequestManager>
+  {
+    if (!TypeUtility.IsDefined (window.Sys))
+      return null;
+    if (!TypeUtility.IsDefined(Sys.WebForms))
+      return null;
+    if (!TypeUtility.IsDefined(Sys.WebForms.PageRequestManager))
+      return null;
+    return Sys.WebForms.PageRequestManager.getInstance();
+  };
+
+  private GetPostBackSettings(): Nullable<Sys.WebForms.PostBackSettings>
+  {
+    var pageRequestManager = this.GetPageRequestManager();
+    if (pageRequestManager == null || !TypeUtility.HasPropertyOfType(pageRequestManager, "_postBackSettings", "object"))
+      return null;
+    return pageRequestManager._postBackSettings as Sys.WebForms.PostBackSettings;
+  }
+
+  private IsSynchronousPostBackRequired(postBackSettings: Nullable<Sys.WebForms.PostBackSettings>): boolean
+  {
+    // TODO: fix in own issue?
+    ArgumentUtility.CheckNotNull ("postBackSettings", postBackSettings);
+
+    if (postBackSettings.async == false)
+      return true;
+
+    return Array.contains(this._synchronousPostBackCommands, postBackSettings.asyncTarget + '|' + this._theForm.__EVENTARGUMENT.value);
+  };
+
+  public SetIsSubmitting (isAsynchronous: boolean, eventTarget: string, eventArgument: string): SubmitState
+  {
+    var submitterElement = this.GetSubmitterOrActiveElement();
+    if (submitterElement != null)
+    {
+      $(submitterElement).addClass('SmartPageSubmitter');
+    }
+
+    $('html').addClass('SmartPageBusy');
+
+    var isAutoPostback = false;
+    if (submitterElement != null)
+    {
+      var tagName = submitterElement.tagName.toLowerCase();
+      if (tagName === 'input')
+      {
+        var type = (submitterElement as HTMLInputElement).type.toLowerCase();
+        isAutoPostback = type !== 'submit' && type !== 'button';
+      }
+      else if (tagName === 'textarea')
+      {
+        isAutoPostback = true;
+      }
+      else if (tagName === 'select')
+      {
+        isAutoPostback = true;
+      }
+    }
+
+    var submitState: SubmitState = {
+      CancelSubmit: false,
+      IsAsynchronous: isAsynchronous,
+      IsAutoPostback: isAutoPostback,
+      Submitter: submitterElement,
+      EventTarget: eventTarget,
+      EventArgument: eventArgument,
+      NextSubmitState : null
+    };
+    if (this._submitState == null)
+    {
+      this._submitState = submitState;
+    }
+    else if (this._submitState.NextSubmitState == null)
+    {
+      if (this._submitState.IsAutoPostback && !isAutoPostback)
+      {
+        this._submitState.NextSubmitState = submitState;
+      }
+      else
+      {
+        // TODO: the other properties are undefined and not null
+        this._submitState.NextSubmitState = { CancelSubmit: true, Submitter: null } as any;
```

Comment:
This was a poor man's version of 
Type = {prop1=true, prop2} | {prop1=false, prop3}

What's called again?

Eitherway, we should rewrite this now that we're typesafe.

Reply:
discriminated union? If this is the case what is the discriminator? `CancelSubmit`? Because TypeScript does support discriminated unions

Reply:
Yes, DU, thanks! And CancelSubmit it is.

Reply:
Actually, right now it's Submitter != null as the discriminator. CancelSubmit is just to mark what we intend because this property is presently unsed after initialization. we should do a DU with Submitter!=null first and then consider a follow-up task if we feel it's helpful.

Reply:
I re-read the code. Submitter!=null is the disciminator most likely. but it's actually a bit more complicated since CancelSubmit=false can also end up with a Submitter=null. And we don't test for CancelSubmit, just for Submitter!=null. So, that's a follow-up task I'm afraid :/

Reply:
Yep, re-checked the code, Submitter!=null covers all discrimation points. but the intialization will need updating because right now, Submitter is just one of many members but all the other members will be effectively ignored once we test for Submitter. The changes to support a DU will mean this has to be a separate ticket. But it should be a fun task to do this correctly. Last time we touched this, we introduced a really great bug that broke the world.